src/Halogen/HTML.purs

```diff
@@ -67,4 +70,53 @@ slot
   -> (o -> Maybe act)
   -> ComponentHTML' act ps m
 slot sym p component input outputQuery =
-  Core.slot (componentSlot sym p component input outputQuery)
+  Core.slot (ComponentSlot (componentSlot sym p component input outputQuery))
+
+-- | Optimizes rendering of a subtree given an equality predicate. If an argument
+-- | is deemed equivalent to the previous value, rendering and diffing will be
+-- | skipped. You should not use this function fully saturated, but instead
+-- | partially apply it for use within a Component's scope. For example, to skip
+-- | rendering for equal states, just wrap your `render` function.
+-- |
+-- | ```purescript
+-- | myComponent = component
+-- |  { render: memoized eq render
+-- |  , ...
+-- |  }
+-- | ```
+memoized
+  :: forall a act ps m
+   . (a -> a -> Boolean)
+  -> (a -> ComponentHTML' act ps m)
+  -> a
+  -> ComponentHTML' act ps m
+memoized eqFn f a = Core.slot (ThunkSlot (thunked eqFn f a))
+
+-- | Skips rendering for referentially equal arguments. You should not use this
+-- | function fully saturated, but instead partially apply it for use within a
+-- | Component's scope.
+lazy
+  :: forall a act ps m
+   . (a -> ComponentHTML' act ps m)
+  -> a
```

Comment:
don't you need Eq instance for it?

Reply:
I thought the "referentially" was related to "referential transparency", what about `Skips rendering for arguments which are equal by reference`?